File: 01-machine/01-memory-locality/python/locality.py
"""
Layer 1 - Memory & cache locality
Pointer-chasing benchmark: same logical traversal (visit N nodes exactly once
each lap), same amount of "work" (one addition per step), but two different
physical memory layouts:

  sequential -> node i's successor lives at i+1 in memory (cache-friendly)
  shuffled   -> node i's successor is a random other node (cache-hostile)

If locality mattered as much as the roadmap claims, shuffled should be
noticeably slower than sequential even though both do the exact same number
of additions. Run it and see what Python's interpreter overhead does to that
signal.
"""
import random
import time

N = 2_000_000        # big enough to blow past L2/L3 cache
LAPS = 5              # full traversals of all N nodes


def build(shuffled: bool):
    values = list(range(N))
    next_idx = [0] * N
    if not shuffled:
        for i in range(N):
            next_idx[i] = (i + 1) % N
    else:
        perm = list(range(N))
        random.shuffle(perm)
        for i in range(N):
            next_idx[perm[i]] = perm[(i + 1) % N]
    return values, next_idx


def traverse(values, next_idx, laps):
    total = 0
    idx = 0
    steps = N * laps
    for _ in range(steps):
        total += values[idx]
        idx = next_idx[idx]
    return total


def bench(label, shuffled):
    values, next_idx = build(shuffled)
    start = time.perf_counter()
    total = traverse(values, next_idx, LAPS)
    elapsed = time.perf_counter() - start
    ns_per_step = elapsed / (N * LAPS) * 1e9
    print(f"{label:10s}  total={total:>15d}  time={elapsed:6.3f}s  {ns_per_step:6.1f} ns/step")


if __name__ == "__main__":
    print(f"N={N:,} laps={LAPS} (python {__import__('platform').python_version()})")
    bench("sequential", shuffled=False)
    bench("shuffled", shuffled=True)

File: 01-machine/01-memory-locality/cpp/locality.cpp
// Layer 1 - Memory & cache locality, C++ version.
// Same pointer-chasing benchmark as the other languages: two physical
// layouts of the same logical traversal. Compiled with optimizations on,
// this is the closest thing in the lab to "what the hardware actually
// does" with none of a managed runtime's overhead in the way -- which is
// exactly why the shuffled/sequential ratio here should be the widest of
// any language in this lab (see the Python version's README note on how
// interpreter overhead dilutes this same signal).
#include <chrono>
#include <cstdint>
#include <cstdio>
#include <random>
#include <vector>

constexpr int64_t N = 2'000'000;
constexpr int64_t LAPS = 5;

void build(bool shuffled, std::vector<int32_t>& values, std::vector<int32_t>& next) {
    values.resize(N);
    next.resize(N);
    for (int64_t i = 0; i < N; i++) values[i] = static_cast<int32_t>(i);

    if (!shuffled) {
        for (int64_t i = 0; i < N; i++) next[i] = static_cast<int32_t>((i + 1) % N);
        return;
    }
    std::vector<int64_t> perm(N);
    for (int64_t i = 0; i < N; i++) perm[i] = i;
    std::mt19937_64 rng(42);
    for (int64_t i = N - 1; i > 0; i--) {
        std::uniform_int_distribution<int64_t> dist(0, i);
        int64_t j = dist(rng);
        std::swap(perm[i], perm[j]);
    }
    for (int64_t i = 0; i < N; i++) {
        next[perm[i]] = static_cast<int32_t>(perm[(i + 1) % N]);
    }
}

int64_t traverse(const std::vector<int32_t>& values, const std::vector<int32_t>& next, int64_t laps) {
    int64_t total = 0;
    int32_t idx = 0;
    int64_t steps = N * laps;
    for (int64_t s = 0; s < steps; s++) {
        total += values[idx];
        idx = next[idx];
    }
    return total;
}

void bench(const char* label, bool shuffled) {
    std::vector<int32_t> values, next;
    build(shuffled, values, next);
    auto start = std::chrono::high_resolution_clock::now();
    int64_t total = traverse(values, next, LAPS);
    auto end = std::chrono::high_resolution_clock::now();
    double elapsed = std::chrono::duration<double>(end - start).count();
    double ns_per_step = elapsed / (N * LAPS) * 1e9;
    std::printf("%-10s  total=%15lld  time=%6.3fs  %6.1f ns/step\n", label, (long long)total, elapsed, ns_per_step);
}

int main() {
    std::printf("N=%lld laps=%lld (g++/clang++, -O2)\n", (long long)N, (long long)LAPS);
    bench("sequential", false);
    bench("shuffled", true);
    return 0;
}
